### DISCUSSION TITLE
2 classes (server, email) instead of 1

### DISCUSSION THREAD

**Discussion by theomantadelis** [PHPMailer/PHPMailer#2552] [Ideas]
Dear,

I just downloaded PHPMailer, and yes it does what I need. But I cant say it is practical.

I would expect 2 classes:
1) 
class MailServer {
...
}
2)
class Email {
...
}

$email_server = new MailServer("settings...");
$my_email = new Email($to, $from, $subject, $body); // these are the 4 required ones
$my_email->addAddress($to2);
$my_email->addAttachment($file);
etc...
$email_server->send($my_email);

Now for me to use it I will make some encapsulation... But even conceptually an email and email_server are 2 different things. So the should be 2 different classes.

Cheers
ps: if I like my encapsulation I will share it...

**Comment by theomantadelis** [comment]
I do not like it, its ugly but does what I want...
Forgive my naming... php naming seems messing to start with :-D

```php

<?php
//Import PHPMailer classes into the global namespace
//These must be at the top of your script, not inside a function
use PHPMailer\PHPMailer\PHPMailer;
use PHPMailer\PHPMailer\SMTP;
use PHPMailer\PHPMailer\Exception;

//Load Composer's autoloader
require 'vendor/autoload.php';


class MailServerSettings {
    public $SMTPDebug = SMTP::DEBUG_SERVER;
    public $Host = "localhost";
    public $SMTPAuth = false;
    public $Username = "";
    public $Password = "";
    public $SMTPSecure = 'ssl';
    public $Port = 465;
}

class Email {
    public $to_addresses = [];
    public $from_address = "";
    public $from_name = "";
    public $Subject;
    public $Body;
    public $Altbody = "";
    public $Attachments;
    
    function __construct($subject, $body, $to, $name = '') {
        $this->addAddress($to, $name);
        $this->Subject = $subject;
        $this->Body = $body;
        $this->AltBody = $body;
    }

    function addAddress($address, $name = "") {
        $this->to_addresses[$address] = $name;
    }

    function addAttachment($file) {
        $this->Attachments[] = $file;
    }
}

class EmailServer extends PHPMailer {

    function __construct($exceptions = null, $settings) {
        parent::__construct($exceptions);
        $this->isSMTP();
        $settings_vars = get_class_vars(get_class($settings));
        foreach ($settings_vars as $name => $value) {
            $this->$name = $settings->$name;
        }
    }

    function send_email($email) {
        if ($email->from_address != "")
            $this->setFrom($email->from_address, $email->from_name);
        else
            $this->setFrom($this->Username);
        foreach ($email->to_addresses as $to_address => $to_name) {
            $this->addAddress($to_address, $to_name);
        }
        $this->isHTML(true);
        $this->Subject = $email->Subject;
        $this->Body = $email->Body;
        $this->AltBody = $email->AltBody;
        $this->send();
    }

}
?>
// Above code will be used with a require...
<?php
$server = new EmailServer(false, new MailServerSettings());
$email = new Email("Testing", "Everything", "to_someone@somewhere.org");
$server->send_email($email);

?>
```

**Comment by Synchro** [comment]
I recommend SwiftMailer if you want to work that way.

I'd be interested to hear about how you would propose to deal with mail servers that require changes in message formatting that you can't know about until sending time.

**Reply by theomantadelis** [reply]
SwiftMailer forwarded me to Symphony which looks to be close to what I suggested.
But then it would require more and more packages to install in order to support other servers etc...

Anyhow, for some reason I chose PHPMailer and it does what I want so okay I will stick with it...
I already wrote the encapsulation I wanted with 3 classes (settings, email, server) which for the base case makes it really simple to use.

Regarding your comment of servers requiring changes in message formatting. Can you elaborate?
I thought it was pretty standard of how the email looked...

Assuming that you attempt to send an email and the server requests a change, then I would fire an event from the mail server with the email and server as an event argument (and what other info needed like the server response?).
The user must catch the event and reformat the email and send it again.

Now how is it handled?

Cheerios

ps: Thank you for making the code inline... couldnt figure it out my self...

**Comment by Synchro** [comment]
I'm glad you've built a wrapper that works the way you like – that's the point of open source.

The problem area is SMTPUTF8. It requires very different message formatting and encoding, and you can't know whether a server supports it until you connect, so you can't render the message before you connect. One problem is that you can't start out with an SMTPUTF8 message and then re-encode it because some things are simply incompatible – for example there is no way to have header labels use UTF-8 without SMTPUTF8.
SwiftMailer avoids this problem by requiring you to know in advance whether a server supports it or not, and even then it only supports a very small subset of it (only local parts of email addresses). PHPMailer doesn't currently handle it at all, but I've been wondering for several years what the most elegant way to handle it is. It's not simple!
